### PR TITLE
Use always first the local channels

### PR DESCRIPTION
**Pull request by larkox**
#### Summary
Always use autocomplete local channels, and then append the remote channels.

This may cause some jumping of the channels, but that should not be common, since the only channels that will pop up are the channels we are not a member of.

This is related to https://github.com/mattermost/mattermost-mobile/pull/6593/

Also added some fixes here and there, that may be shared with the other linked PR (I will leave the fixes here, and we will see which one merge first).

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-46648

#### Release Note
```release-note
Improve autocomplete channels behaviour
```

### CODE REVIEW

**Review by jfrerich — APPROVED** [review]
LGTM and thanks for the 1:1 walkthrough!

**Review by enahum — APPROVED** [review]

**Comment by enahum** [comment]
@larkox needs conflict resolution

**Comment by larkox** [comment]
Back to dev review, since I added several changes to fix some bugs with corner cases.

**Review by avinashlng1080 — APPROVED** [review]

File: app/components/autocomplete/at_mention/at_mention.tsx [hunk 1]
```diff
@@ -175,9 +175,37 @@ const makeSections = (teamMembers: Array<UserProfile | UserModel>, usersInChanne
     return newSections;
 };
 
+const searchGroups = async (serverUrl: string, matchTerm: string, useGroupMentions: boolean, isChannelConstrained: boolean, isTeamConstrained: boolean, channelId?: string, teamId?: string) => {
```
Comment (enahum) [inline app/components/autocomplete/at_mention/at_mention.tsx:178]: Should we move this one to remote actions?

File: app/components/autocomplete/channel_mention/channel_mention.tsx [hunk 2]
```diff
@@ -306,21 +274,41 @@ const ChannelMention = ({
...
-        runSearch(serverUrl, matchTerm);
-    }, [matchTerm]);
+        setLoading(true);
+        runSearch(serverUrl, matchTerm, teamId);
+    }, [matchTerm, teamId]);
+
+    const channels = useMemo(() => {
+        const ids = new Set(localChannels.map((c) => c.id));
+        return [...localChannels, ...remoteChannels.filter((c) => !ids.has(c.id))].sort((a, b) => {
+            const aDisplay = 'display_name' in a ? a.display_name : a.displayName;
```
Comment (enahum) [inline app/components/autocomplete/channel_mention/channel_mention.tsx:284]: I think there is a sorting function for this already created, does that one qork here?

**Comment by enahum** [comment]
@josephbaylon build ready

**Review by josephbaylon — CHANGES_REQUESTED** [review]
@larkox
1. Android works as expected. At-mention autocomplete now closes once user taps on an option. Also, autocomplete stays open as long as long as it matches a pattern.
2. iOS closes at-mention autocomplete once user taps on an option. However, when user types in characters which should keep the autocomplete open, the autocomplete closes and never shows up again. I couldn't figure out the pattern when it closes but at least on 2 occasions, the autocomplete closes on 2nd character and on third character (see images below).
atmention-2-chars
atmention-3-chars

**Comment by larkox** [comment]
@josephbaylon What you are experiencing seems out of the ordinary. Both apps should work the same.

What I am experiencing in my local is the server not returning the list of users (happens in my local, but not in community). Since users is still using the remote results, the behaviour is "correct" on the client: the server says there are no more users, so I show no more users. I will check if I find why the server is behaving this way.

File: app/components/autocomplete/at_mention/at_mention.tsx [hunk 3]
```diff
@@ -175,9 +175,37 @@ const makeSections = (teamMembers: Array<UserProfile | UserModel>, usersInChanne
...
+    try {
+        if (useGroupMentions && matchTerm && matchTerm !== '') {
+            let g = emptyGroupList;
+
+            // If the channel is constrained, we only show groups for that channel
```
[the remark this answers is not in the document]
Comment (jfrerich) [reply app/components/autocomplete/at_mention/at_mention.tsx:183]: Something like will look less likely to fallthrough and possibly hit multiple if statements.

**Review by jfrerich — APPROVED** [review]
I only have 1 nit after looking at the new code from my last review.

**Comment by josephbaylon** [comment]
@larkox Did anything change? I don't think the issue I posted above has been resolved yet.

**Comment by larkox** [comment]
@josephbaylon Sorry! I thought the issue was only on the server side, because what I was receiving was different the last time I tested it, but now I am seeing what you described. Taking a look right now.

**Comment by larkox** [comment]
@josephbaylon Now should be ready. One piece of code apparently unneeded was creating a race condition.

**Review by josephbaylon — APPROVED** [review]
LGTM. Issues found are now fixed. Also PR fixes issue of autocomplete not getting closed after user taps on item.

**Pull request merged**